cocos/core/event/eventful.ts

```diff
@@ -0,0 +1,198 @@
+     * }, node);
+     */
+    once<TFunction extends Function> (type: EventType, callback: TFunction, thisArg?: any): typeof callback;
+
+    /**
+     * @en
+     * Removes the listeners previously registered with the same type, callback, target and or useCapture,
+     * if only type is passed as parameter, all listeners registered with that type will be removed.
+     * @zh
+     * 删除之前用同类型，回调，目标或 useCapture 注册的事件监听器，如果只传递 type，将会删除 type 类型的所有事件监听器。
+     *
+     * @param type - A string representing the event type being removed.
+     * @param callback - The callback to remove.
+     * @param target - The target (this object) to invoke the callback, if it's not given, only callback without target will be removed
+     * @example
+     * // register fire eventListener
+     * var callback = eventTarget.on('fire', function () {
+     *     cc.log("fire in the hole");
+     * }, target);
+     * // remove fire event listener
+     * eventTarget.off('fire', callback, target);
+     * // remove all fire event listeners
+     * eventTarget.off('fire');
+     */
+    off<TFunction extends Function> (type: EventType, callback?: TFunction, thisArg?: any): void;
+
+    /**
+     * @en Removes all callbacks previously registered with the same target (passed as parameter).
+     * This is not for removing all listeners in the current event target,
+     * and this is not for removing all listeners the target parameter have registered.
+     * It's only for removing all listeners (callback and target couple) registered on the current event target by the target parameter.
+     * @zh 在当前 EventTarget 上删除指定目标（target 参数）注册的所有事件监听器。
+     * 这个函数无法删除当前 EventTarget 的所有事件监听器，也无法删除 target 参数所注册的所有事件监听器。
+     * 这个函数只能删除 target 参数在当前 EventTarget 上注册的所有事件监听器。
+     * @param typeOrTarget - The target to be searched for all related listeners
+     */
+    targetOff (typeOrTarget: string | object): void;
+
+    /**
+     * @zh 移除在特定事件类型中注册的所有回调或在某个目标中注册的所有回调。
+     * @en Removes all callbacks registered in a certain event type or all callbacks registered with a certain target
+     * @param typeOrTarget - The event type or target with which the listeners will be removed
+     */
+    removeAll (typeOrTarget: string | object): void;
+
+    /**
+     * @zh 派发一个指定事件，并传递需要的参数
+     * @en Trigger an event directly with the event name and necessary arguments.
+     * @param type - event type
+     * @param args - Arguments when the event triggered
+     */
+    emit (type: EventType, ...args: any[]): void;
+}
+
+/**
+ * 生成一个类，该类继承自指定的基类，并实现了 `IEventful` 的所有接口。
+ * @param base 基类。
+ * @example
+ * ```ts
+ * class Base { say() { console.log('Hello!'); } }
+ * class MyClass extends Eventful(Base) { }
+ * function (o: MyClass) {
+ *     o.say(); // Ok: 继承自 `Base`
+ *     o.emit('sing', 'The ghost'); // Ok: `MyClass`类的对象实现了 IEventful
+ * }
+ * ```
+ */
+export function Eventful<TBase> (base: Constructor<TBase>): Constructor<TBase & IEventful> {
```

Comment:
I don't think Eventful is ideal, I propose 

IEventful -> IEventTarget
Eventful -> EventTargetExtended
EventTarget = EventTargetExtended(Object)

Make things clearer

Reply:
`EventTargetExtended` is so wordy